src/crimson/osd/pg_backend.cc

```diff
@@ -830,6 +830,107 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
     return get_attr_errorator::now();
   });
 }
+static inline int do_xattr_cmp_str(int op, string& v1s, bufferlist& xattr)
+{
+  string v2s(xattr.c_str(), xattr.length());
+
+  logger().debug("do_xattr_cmp_str '{}' vs '{}' op {}", v1s, v2s, op);
+
+  switch (op) {
+  case CEPH_OSD_CMPXATTR_OP_EQ:
+    return (v1s.compare(v2s) == 0);
+  case CEPH_OSD_CMPXATTR_OP_NE:
+    return (v1s.compare(v2s) != 0);
+  case CEPH_OSD_CMPXATTR_OP_GT:
+    return (v1s.compare(v2s) > 0);
+  case CEPH_OSD_CMPXATTR_OP_GTE:
+    return (v1s.compare(v2s) >= 0);
+  case CEPH_OSD_CMPXATTR_OP_LT:
+    return (v1s.compare(v2s) < 0);
+  case CEPH_OSD_CMPXATTR_OP_LTE:
+    return (v1s.compare(v2s) <= 0);
+  default:
+    return -EINVAL;
+  }
+}
+
+static inline int do_xattr_cmp_u64(int op, __u64 v1, bufferlist& xattr)
+{
+  __u64 v2;
+
+  string v2s(xattr.c_str(), xattr.length());
+  if (v2s.length())
+    v2 = strtoull(v2s.c_str(), NULL, 10);
+  else
+    v2 = 0;
+
+  logger().debug("do_xattr_cmp_u64 '{}' vs '{}' op {}", v1, v2, op);
+
+  switch (op) {
+  case CEPH_OSD_CMPXATTR_OP_EQ:
+    return (v1 == v2);
+  case CEPH_OSD_CMPXATTR_OP_NE:
+    return (v1 != v2);
+  case CEPH_OSD_CMPXATTR_OP_GT:
+    return (v1 > v2);
+  case CEPH_OSD_CMPXATTR_OP_GTE:
+    return (v1 >= v2);
+  case CEPH_OSD_CMPXATTR_OP_LT:
+    return (v1 < v2);
+  case CEPH_OSD_CMPXATTR_OP_LTE:
+    return (v1 <= v2);
+  default:
+    return -EINVAL;
+  }
+}
+PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
+  const ObjectState& os,
+  OSDOp& osd_op) const
+{
+  std::string name;
+  auto bp = osd_op.indata.cbegin();
+  {
+    std::string aname;
+    bp.copy(osd_op.op.xattr.name_len, aname);
+    name = "_" + aname;
+    name[osd_op.op.xattr.name_len + 1] = 0;
+  }
+  logger().debug("cmpxattr on obj={} for attr={}", os.oi.soid, name);
+  return getxattr(os.oi.soid, name).safe_then_interruptible(
+    [&osd_op, name] (auto &&xattr) {
+    int result = 0;
+    auto bp = osd_op.indata.cbegin();
+    switch (osd_op.op.xattr.cmp_mode) {
+    case CEPH_OSD_CMPXATTR_MODE_STRING:
+      {
+        string val;
+        bp.copy(osd_op.op.xattr.value_len, val);
+	val[osd_op.op.xattr.value_len + 1] = 0;
+        result = do_xattr_cmp_str(osd_op.op.xattr.cmp_op, val, xattr);
+      }
+    break;
+    case CEPH_OSD_CMPXATTR_MODE_U64:
+      {
+        uint64_t u64val;
+        try {
+          decode(u64val, bp);
+	} catch (ceph::buffer::error& e) {
+          throw crimson::osd::invalid_argument();
+	}
+        result = do_xattr_cmp_u64(osd_op.op.xattr.cmp_op, u64val, xattr);
+      }
+    break;
+    default:
+        logger().debug("bad cmp mode {}", osd_op.op.xattr.cmp_mode);
+        result = -EINVAL;
+    }
+    if (!result) {
+      result = -ECANCELED;
+    }
+    osd_op.rval = result;
```

Comment:
```suggestion
    if (result == 0) {
      logger().info("cmp_xattr: comparison returned false")
      osd_op.rval = -ECANCELED;
    } else {
      osd_op.rval = result;
    }
```

because, it seems like an unexpected result, so we might want to log this as we do in classic osd.